File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/pagination/Page.java
package in.org.quicko.silo.client.pagination;

import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import java.util.OptionalInt;

/**
 * The shared shape of one page of results: the rows, the answered window and the
 * total. It loads nothing — a concrete page declares its own {@code next()} and
 * {@code previous()} returning its own type.
 *
 * <p>{@link #truncated()} marks a page whose {@code total} counts what a scan
 * examined rather than what exists. {@link #pageCount()} is then unknowable and
 * {@link #hasMore()} falls back to "this page came back full".
 */
public abstract class Page<Row> implements Iterable<Row> {
  private final List<Row> rows;
  private final PageWindow window;
  private final int total;
  private final boolean truncated;

  protected Page(List<Row> rows, int total, PageWindow window) {
    this(rows, total, window, false);
  }

  protected Page(List<Row> rows, int total, PageWindow window, boolean truncated) {
    this.rows = List.copyOf(rows);
    this.total = total;
    this.window = window;
    this.truncated = truncated;
  }

  /** How many rows match, across every page. */
  public int total() {
    return total;
  }

  public boolean truncated() {
    return truncated;
  }

  public int limit() {
    return window.limit();
  }

  public int offset() {
    return window.offset();
  }

  public int pageNumber() {
    return Math.floorDiv(offset(), limit()) + 1;
  }

  /** Empty when the total is a scan's count rather than a real one. */
  public OptionalInt pageCount() {
    if (truncated) return OptionalInt.empty();
    return OptionalInt.of(Math.max(1, (int) Math.ceil((double) total / limit())));
  }

  public boolean hasMore() {
    if (truncated) return rows.size() == limit();
    return offset() + rows.size() < total;
  }

  public boolean isEmpty() {
    return rows.isEmpty();
  }

  public int size() {
    return rows.size();
  }

  protected List<Row> rows() {
    return rows;
  }

  protected PageWindow window() {
    return window;
  }

  /** The window the next page would load, so no subclass repeats the
   *  "only if there is more" arithmetic. */
  protected Optional<PageWindow> windowForNext() {
    return hasMore() ? Optional.of(window.next()) : Optional.empty();
  }

  @Override
  public Iterator<Row> iterator() {
    return rows.iterator();
  }
}

File: packages/silo-client-dart/lib/src/pagination/page.dart
import 'dart:math';

import 'package:meta/meta.dart';

import 'page_window.dart';

/// One page of results. [truncated] marks a [total] that counts what a scan
/// examined: [pageCount] is then null and [hasMore] means "this page is full".
abstract class Page<Row> extends Iterable<Row> {
  Page(this._rows, this.total, this._window, {this.truncated = false});

  final List<Row> _rows;
  final PageWindow _window;
  final int total;
  final bool truncated;

  int get limit => _window.limit;

  int get offset => _window.offset;

  int get pageNumber => offset ~/ limit + 1;

  int? get pageCount => truncated ? null : max(1, (total / limit).ceil());

  bool get hasMore => truncated ? _rows.length == limit : offset + _rows.length < total;

  @override
  Iterator<Row> get iterator => _rows.iterator;

  @override
  int get length => _rows.length;

  @protected
  PageWindow? windowForNext() => hasMore ? _window.next() : null;

  @protected
  PageWindow? windowForPrevious() => _window.previous();
}
